lib/src/conflicts.rs

```diff
@@ -820,4 +849,25 @@ mod tests {
         assert_eq!(c(&[-1], &[4, 9]).try_map(sqrt), Err(()));
         assert_eq!(c(&[1], &[-4, 9]).try_map(sqrt), Err(()));
     }
+
+    #[test]
+    fn test_flatten() {
+        fn c<T: Clone>(removes: &[T], adds: &[T]) -> Conflict<T> {
+            Conflict::new(removes.to_vec(), adds.to_vec())
+        }
+        // 1-way conflict of 1-way conflict
+        assert_eq!(c(&[], &[c(&[], &[0])]).flatten(), c(&[], &[0]));
+        // 1-way conflict of 3-way conflict
+        assert_eq!(c(&[], &[c(&[0], &[1, 2])]).flatten(), c(&[0], &[1, 2]));
+        // 3-way conflict of 1-way conflicts
+        assert_eq!(
+            c(&[c(&[], &[0])], &[c(&[], &[1]), c(&[], &[2])]).flatten(),
+            c(&[0], &[1, 2])
+        );
+        // 3-way conflict of 3-way conflicts
+        assert_eq!(
+            c(&[c(&[0], &[1, 2])], &[c(&[3], &[4, 5]), c(&[6], &[7, 8])]).flatten(),
+            c(&[1, 2, 3, 6], &[4, 7, 0, 5, 8])
+        );
```

Comment:
Noob question: Does the order of flattened conflict matter?

Suppose conflict is a tree that looks like
```
A0   A1    A2
 \  /  \  /
  R0    R1
```
I feel flattened nodes would naturally be in `A0, R0^-1, A1, R1^-1, A2` order.

So 3-way/3-way conflicts will be
```
4 5   7 8
 3     6
   1 2
    0
```
flattened to
```
4 5 0 7 8
 3 1 2 6
```

Reply:
Nice illustration!

I tried to preserve the individual diffs as much as possible. I think it's useful to preserve diffs where possible, so the diffs we present to the user match a diff they might have seen before. We generally interpret a conflict state as a diff from the missing state to the first add, then a diff between each consecutive pair. Your method also mostly preserves diffs. I suppose we could preserve diffs and preserve the order between diffs by tweaking your method to move the first element last when adding a negative term. Like this:


```
Add A0:
4 5
 3

Add R0 (with the '1' moved last):
4 5 0
 3 2 1

Add A1:
4 5 0 7 8
 3 2 1 6
```

That way we also preserve the `0->2` diff (but reversed since it was in a negative position). What do you think? I'm happy to make that change.

Reply:
I've made that change. The code turned out simpler too. Thanks for the suggestion! And let me know if you disagree about moving the first add last in removes.

Reply:
New code indeed looks better. Thanks for the explanation.